### ISSUE TITLE
No ability for multiple entities to reference the same material entity

### ISSUE THREAD

**Issue by Aitolda** [kasenvr/project-athena#114]
At the moment only a single entity can be influenced by a material entity, requiring a separate material entity from every mesh that want's to use this material.  If we were able to allow multiple entities dramatically simplifies workflow for creators.  Right now the connection is made by turning a material entity into the child of the mesh it influences.  Perhaps there is another way to connect these allowing for flexibility.

**Comment by HifiExperiments** [comment]
the trick with this is that we made material entities follow their parent so you wouldn’t end up with random material entities all over your domain, and so that we only load material entities as needed (when you see them or their parent)

perhaps a solution here is a material library?  each material entity still only affects one entity, but you have some way of defining a material once and then reusing it, and having changes to that one material affect all instances...

**Comment by Aitolda** [comment]
"perhaps a solution here is a material library? each material entity still only affects one entity, but you have some way of defining a material once and then reusing it, and having changes to that one material affect all instances"

This sounds like a reasonable alternative to me, of course, but I must confess my ignorance of the level of work needed.

**Comment by SilverfishVR** [comment]
Tl:DR: I advocate careful consideration over speedy implementation.

It is important to distinguish between material entities, and material resources.
Material resources are extremely useful because they let you control and fine-tune each material parameter, in-app, in-location, real-time, circumventing all of the import-export issues, the problem is, that the only way to apply a material from a resource is through a material entity, this is very impractical due to the parenting system where a material can only be applied (parented) to 1 UUID and the fact that you can’t, by selecting model entity, easily figure out which mesh part is overridden by what material entity.

The goal, I think should be a system that allows me to apply a material to as many entities as I want, and allows me to apply as many materials to an entity as I want(within reason). While maintaining the possibility to “live edit” the material data and to parent entities (for projected materials), all without making the UI an incomprehensible mess.

So far, I think my non-breaking-current-implementation idea is:

Model and shape entities gains the properties currently reserved for material https://apidocs.projectathena.dev/Entities.html#.EntityProperties-Material
With one difference. Model entities needs to support multiple materials so maybe a new type of object? I dunno, but basically:

Model and shape entities can also hold multiple instances of material-entity properties.

This way you can have it all in the entity itself, or reference an external material resource if you want to use it across multiple entities (it could even be one resource with all of your materials).

UI wise, I imagine a “material override” subpage to the _create->Properties_ that  page shows a list of “overrides” each mostly identical to what you currently see in a material entity:
![materialProps](https://user-images.githubusercontent.com/53531160/73122417-12fd1e80-3f85-11ea-8e7f-3300cff27c44.png)

It would also have buttons to Add & remove entries, maybe even a way to copy original model material to override entry.

Q: if a material is applied with different mapping, scale, etc. to multiple entities, can the renderer still treat them as one material?

P.S. parenting materials would still work, it has to if we want to keep projected material mapping (I do) and it could be even more useful now that you can project to entities other than the parent.
Maybe “materialURL” could also be the UUID of a material entity so that when parenting a material entity, it adds a “material override” entry in the parent with a materialURL of its own UUID?

 P.P.S. Currently it seem you can't reference a material by name (as in screenshot) if you use _materialData_  in the materialURL field, but otherwise we could get away with having 1 materialData per entity, rather than 1 per material override.

**Comment by daleglass** [comment]
Dev meeting: PR #1102  may fix this one
